lib/cc-cubic.c

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (c) 2019 Fastly, Janardhan Iyengar
+ * Copyright (c) 2020 RWTH Aachen University, COMSYS Network Architectures Group, Leo Blöcher
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include <math.h>
+#include "quicly/cc.h"
+
+#define QUICLY_MIN_CWND 2
+
+typedef double cubic_float_t;
+#define QUICLY_CUBIC_C ((cubic_float_t)0.4)
+#define QUICLY_CUBIC_BETA ((cubic_float_t)0.7)
+
+/* Calculates the elapsed time since the last congestion event (parameter t) */
+static cubic_float_t calc_cubic_t(const quicly_cc_t *cc)
+{
+    cubic_float_t clock_delta = clock() - cc->state.cubic.avoidance_start;
```

Comment:
I didn't search for an existing clock functionality in quicly before implementing CUBIC. As far as I can see, loss detection is passed the current time via the now parameter, set to `conn->stash.now`. I guess this would be the appropriate time source to use? And its resolution appears to be ms?

Reply:
Correct.

Reply:
Will implement tomorrow.